src/qtt/algorithms/markov_chain.py

```diff
@@ -1,74 +1,88 @@
-# -*- coding: utf-8 -*-
 """ Class to generate signals with continous-time Markov chains
 
 
 @author: pieter.eendebak@gmail.com
 """
 
-# %%
-import numpy as np
+import itertools
 import random
+from dataclasses import dataclass
+from typing import List, Optional, Union
+
+import numpy as np
 import scipy.linalg
-import itertools
 
 
-def _solve_least_squares(a, b):
+def _solve_least_squares(a: np.ndarray, b: np.ndarray) -> np.ndarray:
     rcond = None
     solution = np.linalg.lstsq(a, b, rcond=rcond)[0]
     return solution
 
 
+@dataclass
 class ChoiceGenerator:
-    """ Class to generate random elements with weighted selection """
+    """ Class to generate random elements with weighted selection
 
-    def __init__(self, number_of_states, cum_weights, block_size=5000):
-        """ Class to generate random elements with weighted selection
+    This is a replacement for random.choices that is efficient when a large number of choices has to be generated.
 
-        This is a replacement for random.choices that is efficient when a large number of choices has to be generated.
+    Args:
+        number_of_states: number of choices that has to be generated
+        cum_weights (array[float]): cumulative probabilities of the choices
+        block_size: size of blocks of choices to generate
+    """
 
-        Args:
-            number_of_states (int): number of choices that has to be generated
-            cum_weights (array[float]): cummulative probabilities of the choices
-            block_size (int): size of blocks of choices to generate
-        """
-        if not number_of_states == len(cum_weights):
-            raise Exception('specification of cummulative weights does not match number of states')
+    number_of_states: int
+    cum_weights: np.ndarray
+    block_size: int = 5000
 
-        self.number_of_states = number_of_states
+    def __post_init__(self):
+        if not self.number_of_states == len(self.cum_weights):
+            raise Exception(
+                f'specification of cumulative weights {len(self.cum_weights)} does not match number of states {self.number_of_states}')
         self._idx = 0
```

Comment:
__post_init__ shouldn't add instance attributes (self._block and self._idx). An __init__ should do that or why not add them as (initialized) instance attributes just as block_size.

Reply:
The `_block` and `_idx` are private internal variables. I do not want them in the representation of the object, and also not in the data when the dataclass is converted to json or a dictionary, which is why they are not defined as the `block_size`.
`__init__` is something one should not define for a dataclass